File: go/cmd/tiny-go/session_recovery.go
package main

import (
	"encoding/json"
	"sort"
)

type currentTool struct {
	Name             string `json:"name"`
	DefinitionDigest string `json:"definitionDigest"`
	Replay           string `json:"replay"`
	ReplayKey        string `json:"replayKey"`
}
type currentConfiguration struct {
	ConfigurationDigest string        `json:"configurationDigest"`
	EnvironmentIdentity string        `json:"environmentIdentity"`
	Tools               []currentTool `json:"tools"`
}
type recoveryPlan map[string]any

func currentToolByName(current currentConfiguration, name string) *currentTool {
	for index := range current.Tools {
		if current.Tools[index].Name == name {
			return &current.Tools[index]
		}
	}
	return nil
}

func recoveryAssistant(state sessionState) (string, []any, bool) {
	if state.Operation.Kind != "run" || state.Operation.Step == nil || state.Operation.Step.SettledEntryID == "" {
		return "", nil, false
	}
	for index := len(state.Transcript) - 1; index >= 0; index-- {
		message := state.Transcript[index]
		calls, ok := message["tool_calls"].([]any)
		if message["role"] == "assistant" && ok && len(calls) > 0 {
			return state.Operation.Step.SettledEntryID, calls, true
		}
	}
	return "", nil, false
}

func recoveryCall(call any) (string, string, string) {
	value, _ := call.(map[string]any)
	function, _ := value["function"].(map[string]any)
	id, _ := value["id"].(string)
	name, _ := function["name"].(string)
	arguments, _ := function["arguments"].(string)
	return id, name, arguments
}

func syntheticResult(assistantID string, index int, callID, name, reason string) map[string]any {
	return map[string]any{
		"assistantEntryId": assistantID,
		"toolIndex":        index,
		"toolCallId":       callID,
		"toolName":         name,
		"reason":           reason,
		"content":          syntheticContent[reason],
	}
}

func interruptedResult(tool sessionToolState) map[string]any {
	result := syntheticResult(tool.AssistantEntryID, tool.ToolIndex, tool.ToolCallID, tool.ToolName, "interrupted")
	result["toolStartedId"] = tool.ToolStartedID
	result["resultEntryId"] = tool.ResultEntryID
	return result
}

func planRecovery(state sessionState, current currentConfiguration) recoveryPlan {
	operation := state.Operation
	if operation.Kind == "idle" {
		return recoveryPlan{"type": "finish", "outcome": "completed", "completion": "normal"}
	}

	assistantID, calls, hasAssistant := recoveryAssistant(state)
	pending := []sessionToolState{}
	if operation.Kind == "run" {
		for _, tool := range operation.ToolCalls {
			if tool.Status == "pending" {
				pending = append(pending, tool)
			}
		}
	}
	if operation.AbortRequested {
		if operation.Step != nil && operation.Step.Status == "attempting" {
			return recoveryPlan{"type": "closeAttempt", "error": map[string]any{"code": "aborted", "message": "Operation aborted"}}
		}
		if operation.Kind == "run" && hasAssistant {
			started := map[int]sessionToolState{}
			for _, tool := range operation.ToolCalls {
				if tool.AssistantEntryID == assistantID {
					started[tool.ToolIndex] = tool
				}
			}
			results := []any{}
			for index, raw := range calls {
				callID, name, _ := recoveryCall(raw)
				tool, exists := started[index]
				if exists && tool.Status == "pending" {
					results = append(results, interruptedResult(tool))
				} else if !exists {
					results = append(results, syntheticResult(assistantID, index, callID, name, "aborted"))
				}
			}
			if len(results) > 0 {
				return recoveryPlan{"type": "appendSynthetic", "results": results}
			}
		}
		return recoveryPlan{"type": "finish", "outcome": "aborted"}
	}

	if operation.Step == nil {
		contextID, kind := operation.InputThroughEntryID, operation.Kind
		if operation.Kind == "run" {
			contextID, kind = operation.InputEntryID, "assistant"
		}
		return recoveryPlan{"type": "startStep", "stepKind": kind, "attempt": 1, "contextThroughEntryId": contextID}
	}
	step := operation.Step
	if step.Status == "failed" {
		return recoveryPlan{"type": "finish", "outcome": "failed", "error": map[string]any{"code": "model_error", "message": "provider request failed"}}
	}
	if step.Status == "attempting" {
		if step.Attempt == 2 {
			return recoveryPlan{"type": "blocked", "reason": "attempts_exhausted"}
		}
		if step.ConfigurationDigest != current.ConfigurationDigest {
			return recoveryPlan{"type": "blocked", "reason": "configuration_changed"}
		}
		return recoveryPlan{"type": "startStep", "stepKind": step.StepKind, "attempt": 2, "stepId": step.StepID, "contextThroughEntryId": step.ContextThroughEntryID}
	}
	if operation.Kind == "compaction" {
		return recoveryPlan{"type": "finish", "outcome": "completed", "finalEntryId": operation.ResultEntryID}
	}
	if step.StopReason == "length" && hasAssistant {
		results := make([]any, len(calls))
		for index, raw := range calls {
			callID, name, _ := recoveryCall(raw)
			results[index] = syntheticResult(assistantID, index, callID, name, "truncated")
		}
		return recoveryPlan{"type": "appendSynthetic", "results": results}
	}
	if !hasAssistant {
		return recoveryPlan{"type": "finish", "outcome": "completed", "completion": "normal", "finalEntryId": step.SettledEntryID}
	}

	processed := map[int]bool{}
	for _, tool := range operation.ToolCalls {
		if tool.AssistantEntryID == assistantID {
			processed[tool.ToolIndex] = true
		}
	}
	for index, raw := range calls {
		if processed[index] {
			continue
		}
		callID, name, rawArguments := recoveryCall(raw)
		declaration := currentToolByName(current, name)
		if declaration == nil {
			return recoveryPlan{"type": "appendSynthetic", "results": []any{syntheticResult(assistantID, index, callID, name, "unknownTool")}}
		}
		arguments := map[string]any{}
		if err := json.Unmarshal([]byte(rawArguments), &arguments); err != nil || arguments == nil {
			return recoveryPlan{"type": "appendSynthetic", "results": []any{syntheticResult(assistantID, index, callID, name, "invalidArguments")}}
		}
		return recoveryPlan{"type": "startTool", "mode": "start", "assistantEntryId": assistantID, "toolIndex": index, "toolName": name, "arguments": arguments}
	}

	if len(pending) == 0 {
		last := operation.ToolCalls[len(operation.ToolCalls)-1]
		return recoveryPlan{"type": "startStep", "stepKind": "assistant", "attempt": 1, "contextThroughEntryId": last.ResultEntryID}
	}
	sort.Slice(pending, func(i, j int) bool { return pending[i].ToolIndex < pending[j].ToolIndex })
	tool := pending[0]
	if tool.EnvironmentIdentity != current.EnvironmentIdentity {
		return recoveryPlan{"type": "blocked", "reason": "environment_changed"}
	}
	declaration := currentToolByName(current, tool.ToolName)
	if declaration == nil || declaration.DefinitionDigest == "" {
		return recoveryPlan{"type": "blocked", "reason": "configuration_changed"}
	}
	if tool.Replay == "safe" && declaration.Replay == "safe" && declaration.ReplayKey == tool.ReplayKey {
		return recoveryPlan{"type": "startTool", "mode": "replay", "assistantEntryId": tool.AssistantEntryID, "toolIndex": tool.ToolIndex, "toolStartedId": tool.ToolStartedID, "toolName": tool.ToolName, "arguments": tool.Arguments}
	}
	if tool.Replay == "safe" && (declaration.Replay != "safe" || declaration.ReplayKey != tool.ReplayKey) {
		return recoveryPlan{"type": "blocked", "reason": "replay_declaration_changed"}
	}
	return recoveryPlan{"type": "appendSynthetic", "results": []any{interruptedResult(tool)}}
}

File: python/tiny_agent/session_recovery.py
from __future__ import annotations

import json
from typing import Any

from .session_reducer import SYNTHETIC_CONTENT


def _synthetic(assistant_id, index, call, reason, tool=None):
    value = {
        "assistantEntryId": assistant_id, "toolIndex": index, "toolCallId": call["id"],
        "toolName": call["function"]["name"], "reason": reason, "content": SYNTHETIC_CONTENT[reason],
    }
    if tool:
        value.update(toolStartedId=tool["toolStartedId"], resultEntryId=tool["resultEntryId"])
    return value


def _assistant(state):
    operation = state["operation"]
    settled = operation.get("step", {}).get("settledEntryId")
    if operation["kind"] != "run" or not settled: return None
    message = next((item for item in reversed(state["transcript"]) if item["role"] == "assistant" and item.get("tool_calls")), None)
    return {"assistantEntryId": settled, "calls": message.get("tool_calls", [])} if message else None


def plan_recovery(state: dict, current: dict) -> dict:
    operation = state["operation"]
    if operation["kind"] == "idle": return {"type": "finish", "outcome": "completed", "completion": "normal"}
    assistant = _assistant(state)
    pending = [tool for tool in operation.get("toolCalls", []) if tool["status"] == "pending"]
    if operation.get("abortRequested"):
        if operation.get("step", {}).get("status") == "attempting":
            return {"type": "closeAttempt", "error": {"code": "aborted", "message": "Operation aborted"}}
        if assistant:
            started = {(tool["assistantEntryId"], tool["toolIndex"]): tool for tool in operation.get("toolCalls", [])}
            results = []
            for index, call in enumerate(assistant["calls"]):
                tool = started.get((assistant["assistantEntryId"], index))
                if tool and tool["status"] == "pending": results.append(_synthetic(assistant["assistantEntryId"], index, call, "interrupted", tool))
                elif not tool: results.append(_synthetic(assistant["assistantEntryId"], index, call, "aborted"))
            if results: return {"type": "appendSynthetic", "results": results}
        return {"type": "finish", "outcome": "aborted"}
    if not operation.get("step"):
        return {"type": "startStep", "stepKind": "assistant" if operation["kind"] == "run" else "compaction", "attempt": 1, "contextThroughEntryId": operation.get("inputEntryId", operation.get("inputThroughEntryId"))}
    step = operation["step"]
    if step["status"] == "failed": return {"type": "finish", "outcome": "failed", "error": {"code": "model_error", "message": "provider request failed"}}
    if step["status"] == "attempting":
        if step["attempt"] == 2: return {"type": "blocked", "reason": "attempts_exhausted"}
        if step["configurationDigest"] != current["configurationDigest"]: return {"type": "blocked", "reason": "configuration_changed"}
        return {"type": "startStep", "stepKind": step["stepKind"], "attempt": 2, "stepId": step["stepId"], "contextThroughEntryId": step["contextThroughEntryId"]}
    if operation["kind"] == "compaction": return {"type": "finish", "outcome": "completed", "finalEntryId": operation["resultEntryId"]}
    if step.get("stopReason") == "length" and assistant:
        processed = {(tool["assistantEntryId"], tool["toolIndex"]) for tool in operation.get("toolCalls", [])}
        results = [_synthetic(assistant["assistantEntryId"], i, call, "truncated") for i, call in enumerate(assistant["calls"]) if (assistant["assistantEntryId"], i) not in processed]
        return {"type": "appendSynthetic", "results": results} if results else {"type": "finish", "outcome": "completed", "completion": "truncated", "finalEntryId": assistant["assistantEntryId"]}
    if not assistant: return {"type": "finish", "outcome": "completed", "completion": "normal", "finalEntryId": step["settledEntryId"]}
    processed = {(tool["assistantEntryId"], tool["toolIndex"]) for tool in operation.get("toolCalls", [])}
    for index, call in enumerate(assistant["calls"]):
        if (assistant["assistantEntryId"], index) in processed: continue
        declaration = next((tool for tool in current["tools"] if tool["name"] == call["function"]["name"]), None)
        if not declaration: return {"type": "appendSynthetic", "results": [_synthetic(assistant["assistantEntryId"], index, call, "unknownTool")]}
        try: args = json.loads(call["function"]["arguments"])
        except (json.JSONDecodeError, TypeError): args = None
        if not isinstance(args, dict): return {"type": "appendSynthetic", "results": [_synthetic(assistant["assistantEntryId"], index, call, "invalidArguments")]}
        return {"type": "startTool", "mode": "start", "assistantEntryId": assistant["assistantEntryId"], "toolIndex": index, "toolName": call["function"]["name"], "arguments": args}
    if not pending:
        return {"type": "startStep", "stepKind": "assistant", "attempt": 1, "contextThroughEntryId": operation["toolCalls"][-1]["resultEntryId"]}
    tool = sorted(pending, key=lambda item: item["toolIndex"])[0]
    if tool["environmentIdentity"] != current["environmentIdentity"]: return {"type": "blocked", "reason": "environment_changed"}
    declaration = next((item for item in current["tools"] if item["name"] == tool["toolName"]), None)
    if not declaration or not declaration["definitionDigest"]: return {"type": "blocked", "reason": "configuration_changed"}
    if tool["replay"] == declaration["replay"] == "safe" and tool["replayKey"] == declaration["replayKey"]:
        return {"type": "startTool", "mode": "replay", "assistantEntryId": tool["assistantEntryId"], "toolIndex": tool["toolIndex"], "toolStartedId": tool["toolStartedId"], "toolName": tool["toolName"], "arguments": tool["arguments"]}
    if tool["replay"] == "safe": return {"type": "blocked", "reason": "replay_declaration_changed"}
    call = assistant["calls"][tool["toolIndex"]]
    return {"type": "appendSynthetic", "results": [_synthetic(tool["assistantEntryId"], tool["toolIndex"], call, "interrupted", tool)]}
